packages/libs/sdk/src/api/api.ts

```diff
@@ -89,65 +37,66 @@ export class API {
+    this.graphApiClient = this.urqlClient;
   }
 
-  private createApolloClient(): ApolloClient<NormalizedCacheObject> {
-    const authLink = setContext(async (_, { headers }) => {
-      return {
-        headers: {
-          ...headers,
-          ...{ 'x-api-key': this.graphApiKey },
-          ...this.additionalHeaders,
-        },
-      };
-    });
-
-    const cacheStructure = new InMemoryCache({
-      possibleTypes: generatedPossibleTypes.possibleTypes,
-      typePolicies: {
-        EVMSchemaType: {
-          // Always merge EVMSchemaType objects
-          // Added because of warning: "Cache data may be lost when replacing the ethereum field of a Query object."
-          merge: true,
-        },
-        NFT: {
-          keyFields: ['contractAddress', 'tokenId'],
-        },
-        Collection: {
-          keyFields: ['address'],
-        },
-        Contract: {
-          keyFields: ['address'],
-        },
-        TokenEvent: {
-          keyFields: ['transactionHash', 'transferIndex'],
-        },
-        Transaction: {
-          keyFields: ['hash'],
-        },
-        TrendingCollection: {
-          keyFields: ['collection', ['address']],
-        },
-        Wallet: {
-          keyFields: ['address'],
-        },
+  private createUrqlClient(): Client {
+    const headers = { ...this.additionalHeaders };
+    if (this.graphApiKey) headers['x-api-key'] = this.graphApiKey;
+    const useNftKey = (data: Data) =>
+      `${data['contractAddress']}:${data['tokenId']}`;
+    const useAddressAsKey = (data: Data) => `${data['address']}`;
+    const useTransactionHashAndIndex = (data: Data) =>
+      `${data['transactionHash']}:${data['transferIndex']}`;
+    const urqlCache = cacheExchange({
+      schema,
+      keys: {
+        EVMSchemaType: () => null, // The entity has no key and no parent entity so effectively won't cache
+        Collection: useAddressAsKey,
+        CollectionOHLCVChart: () => null, // Entities without keys will be embedded directly on the parent entity.
+        Contract: (data) => `${data['address']}`,
+        ERC721NFT: useNftKey,
+        ERC721Collection: useAddressAsKey,
+        ERC1155NFT: useNftKey,
+        ERC1155Collection: useAddressAsKey,
+        NFT: useNftKey,
+        NFTContract: useAddressAsKey,
+        TokenAttribute: () => null,
+        TokenContract: useAddressAsKey,
+        TokenEvent: useTransactionHashAndIndex,
+        TokenMintEvent: useTransactionHashAndIndex,
+        TokenBurnEvent: useTransactionHashAndIndex,
+        TokenSaleEvent: useTransactionHashAndIndex,
+        TokenSwapEvent: useTransactionHashAndIndex,
+        TokenTransferEvent: useTransactionHashAndIndex,
+        TokenUpload: () => null,
+        OpenSeaMetadata: () => null,
+        Transaction: (data) => `${data['hash']}`,
+        TrendingCollection: () => null,
+        Wallet: (data) => `${data['address']}`,
+        WalletNFT: () => null,
```

Comment:
Was having some trouble the WalletNFTs returning results that were just the same NFT again and again for the `nfts.getByWallet` function. `null` means that this will be embedded directly on the parent entity.

[See docs here](https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#custom-keys-and-non-keyable-entities)

Reply:
I mostly tried to stick with what we were previously caching, and set everything else as `null` to be on the safe side. I don't see any equivalent for `possibleTypes` with urql so I had to be pretty verbose with the subtypes.